autotelic/fastify-hmac
# refactor: code tweaks

### Summary
- Modify `validateHMAC` to only accept an optional callback param.
- Add `validateRequests` options to enable preValidation hook.
- Remove `verificationErrorMessage` option.
- Modify `verificationError` to accept function that returns Error.
- Update README and examples.
- Add tests for `index.js`

### Test plan
- Confirm tests pass with 100% coverage
- Run the basic example:
```shell
npm run example:basic
```
- In another terminal session make the following request using `httpie`
```sh
http POST localhost:3000/foo Signature:'keyId="test-key-a", algorithm="hs2019", headers="(request-target) (created) (expires) host", signature="fKj7/7DbxLi1RkMqBrPB6HJ6dXUThRM4nwSErMo3ni1SuXZjTEpNDiZozUMUYqpcGssakLQl2y8ryBz0U5oxlg==", created=1402170695, expires=1402170895'
```
- The response should resemble the following:
```sh
HTTP/1.1 200 OK
Connection: keep-alive
Date: Tue, 15 Sep 2020 20:03:28 GMT
content-length: 13
content-type: application/json; charset=utf-8

{
    "foo": "bar"
}
```
- Try modifying the request signature and making another request
- The response should resemble the following:
```sh
HTTP/1.1 401 Unauthorized
Connection: keep-alive
Date: Tue, 15 Sep 2020 20:08:13 GMT
content-length: 83
content-type: application/json; charset=utf-8

{
    "error": "Unauthorized",
    "message": "Signature verification failed",
    "statusCode": 401
}
```
